### DISCUSSION TITLE
Testing

### DISCUSSION THREAD

**Discussion by profOnno** [helge17/tuxguitar#303] [General]
How can I implement testing capabilities?

Now I'm using this guide [https://www.baeldung.com/junit-5](https://www.baeldung.com/junit-5).
[https://www.baeldung.com/maven-cant-find-junit-tests]([https://www.baeldung.com/maven-cant-find-junit-tests])

Using this guide, I have something working.
This also needs to have another tree structure. `src` -> `src/main/java`. The test is in the src folder `src/test/java`, the resources in `src/test/resources`. For maven to pick up on this the pom file needs to be changed:
```xml
<build>
  <sourceDirectory>src/main/java</sourceDirectory>
</build>
```
And I added (this have to go to the main pom?):
```xml
<dependency>
  <groupId>org.junit.jupiter</groupId>
  <artifactId>junit-jupiter-engine</artifactId>
  <version>5.9.2</version>
  <scope>test</scope>
</dependency>
```

Result maven build will fail if test fails. The test I added converts a `.tg file` to `.musicxml` and then runs `xmllint` to validate against the MusicXml-4.0 schema.

**Comment by guiv42** [comment]
Thanks for this information. This is something I would really like to put in place, but I never found the time to learn how (way too many things to learn already...).
I have already implemented a few tests, but as I didn't want to mess up with the rest of the code or to add dependencies I isolated them in TuxGuitar-test module. I would really like to refactor it, and to redistribute the different tests in the modules corresponding to the tested code.
So if you have something currently running, I am highly interested. Do you have something I could look at in one branch of your repo?

**Reply by profOnno** [reply]
Should be this [commit](https://github.com/profOnno/tuxguitar/commit/5335aca4f497edd4f2e6a25460802b06f24c8b2b)

**Comment by guiv42** [comment]
Just for information: automated testing is now in place in tuxguitar-next branch.
Take care, in a Debian environments test methods are only executed if their name starts with "test". Annotation `@Test` is not sufficient.
